src/python/pants/backend/common/targets/app.py

```diff
@@ -0,0 +1,259 @@
+    :param mapper: Function that takes a path string and returns a path string. Takes a path in
+      the source tree, returns a path to use in the resulting bundle. By default, an identity
+      mapper.
+    :param string relative_to: Set up a simple mapping from source path to bundle path.
+    :param fileset: The set of files to include in the bundle.  A string filename, or list of
+      filenames, or a Fileset object (e.g. globs()).
+      E.g., ``relative_to='common'`` removes that prefix from all files in the application bundle.
+    """
+
+    if fileset is None:
+      raise ValueError("In {}:\n  Bare bundle() declarations without a `fileset=` parameter "
+                       "are no longer supported.".format(self._parse_context.rel_path))
+
+    if mapper and relative_to:
+      raise ValueError("Must specify exactly one of 'mapper' or 'relative_to'")
+
+    if rel_path and isinstance(fileset, FilesetWithSpec) and fileset.rel_root != rel_path:
+      raise ValueError("Must not use a glob for 'fileset' with 'rel_path'."
+                       " Globs are eagerly evaluated and ignore 'rel_path'.")
+
+    # A fileset is either a glob, a string or a list of strings.
+    if isinstance(fileset, FilesetWithSpec):
+      pass
+    elif isinstance(fileset, six.string_types):
+      fileset = [fileset]
+    else:
+      fileset = assert_list(fileset, key_arg='fileset')
+
+    real_rel_path = rel_path or self._parse_context.rel_path
+
+    if relative_to:
+      base = os.path.join(get_buildroot(), real_rel_path, relative_to)
+      mapper = RelativeToMapper(base)
+    else:
+      mapper = mapper or RelativeToMapper(os.path.join(get_buildroot(), real_rel_path))
+
+    return BundleProps(real_rel_path, mapper, fileset)
+
+  def create_bundle_props(self, bundle):
+    rel_path = getattr(bundle, 'rel_path', None)
+    mapper = getattr(bundle, 'mapper', None)
+    relative_to = getattr(bundle, 'relative_to', None)
+    fileset = getattr(bundle, 'fileset', None)
+    return self(rel_path, mapper, relative_to, fileset)
+
+
+class BundleField(tuple, PayloadField):
+  """A tuple subclass that mixes in PayloadField.
+
+  Must be initialized with an iterable of Bundle instances.
+  """
+
+  @staticmethod
+  def _hash_bundle(bundle):
+    hasher = sha1()
+    hasher.update(bundle.rel_path)
+    for abs_path in sorted(bundle.filemap.keys()):
+      buildroot_relative_path = os.path.relpath(abs_path, get_buildroot())
+      hasher.update(buildroot_relative_path)
+      hasher.update(bundle.filemap[abs_path])
+      if os.path.isfile(abs_path):
+        # Update with any additional string to differentiate empty file with non-existing file.
+        hasher.update('e')
+        with open(abs_path, 'rb') as f:
+          hasher.update(f.read())
+    return hasher.hexdigest()
+
+  def _compute_fingerprint(self):
+    return combine_hashes(map(BundleField._hash_bundle, self))
+
+
+class App(Target):
```

Comment:
How about `AppBase`, to emphasize that you're expected to subclass it.

The docstring should also be explicit about that. E.g., `A base class for deployable application targets.` or something.